docs/DesignDocs/SGX2-EDMM.md

```diff
@@ -0,0 +1,105 @@
+instructions restrict or extend page permissions, respectively. Page type
+changes (including page removal/trimming) are performed by the `ENCLS[EMODT]`
+instruction.
+
+## Heap allocation
+
+We propose to add a new OCALL to OE enclaves which requests new memory pages
+from the SGX driver (via `ioctl`), combined with an enclave function that, upon
+return from the OCALL, accepts the new pages. This does not require handling of
+signals or page faults and is easily integrated with memory managers like the
+default dlmalloc, by pre-mapping a configurable maximum number of pages and
+tracking the number of currently mapped pages. For performance reasons, we
+propose to add a switchless version of the OCALL whenever
+`OE_ENCLAVE_SETTING_CONTEXT_SWITCHLESS` is enabled:
+
+    untrusted {
+      public oe_result_t oe_add_pages(size_t count);
+      public oe_result_t oe_add_pages_switchless(size_t count) transition_using_threads;
+    };
+
+The pluggable allocator API then uses `oe_add_pages`, before allocating
+additional pages, if necessary, before calling into `oe_allocator_*alloc`. The
+maximum number of heap pages (and perhaps other parameters like page
+permissions) are passed to `oe_allocator_init`.
+
+Applications that run a full OS kernel, for instance SGX-LKL, need to know about
+the size and layout of their address space and can therefore effectively use
+dynamic region allocation, which the SGX driver supports by tracking memory
+regions in the SGX driver. It automatically grows the heap by allocating a
+configurable number of pages upon request for a particular (unmapped) page.
+Dynamic regions are most conveniently declared as an initialization parameter to
+`oe_allocator_init` again.
+
+The permissions (R/W/X) of pages can be modified dynamically. Permissions of
+mapped pages are restricted by the `ENCLS[EMODPR]` instruction, which is a
+privileged instruction (ring 0) that the host executes. Similarly, permissions
+are extended by the `ENCLU[EMODPE]` instruction, which the enclave executes.
+OE's current pluggable allocators do not support permission changes. We propose
+to add a single, easy to use function to change (both restrict and extend) the
+page types of allocated memory, e.g.
+
+    oe_result_t oe_sgx_set_page_permissions(void* address, bool R, bool W, bool X);
+
+We expect that OE applications will usually want the R bit enabled by default,
+but some applications require the X bit to be set and others may chose to
+relinquish the W flag to avoid accidental modification of memory pages. (One of
+the reasons that SGX-LKL is forced to use a custom version of OE, is that the
+memory permissions are not configurable at signing, launching, or run time; see
+also [#3561](https://github.com/openenclave/openenclave/pull/3561).)
+
+## Extensions
+
+SGX2 supports "lazy" addition of heap and stack memory by triggering and
+catching page faults. However, we believe that most near-term applications will
```

Comment:
@CodeMonkeyLeet - I agree with your assessment. I think that we can break the task down into a number of subtasks (which do focus on the virtual memory management).  Here is my list:

1. Kernel APIs
   1. Linux : not supported in kernel. Need to make a proposal for support (and work on the assumption that the final solution will be close to the proposal)
2. Enclave Common Loader: abstracts kernel APIs into a single set of APIs
3. Core Design - the basic memory management:
   1. Enclave Memory Mapping Structures: the enclave needs to know its own memory in order to know what it should and should not EACCEPT/EACCEPTCOPY
        1. Virtual Address Structures - enclave needs to know properties of each memory address region (similar to VMAs)
            1. RWX
            2. Page Type
            3. Properties (e.g GROWS_DOWN/GROWS_UP)
        2. Physical Page - enclave needs to keep track of information such as
            1. Pages which are EACCEPT'ed, TRIM'ed, etc.
    2. Low Level Functions (within the enclave):
        1. Primitives: Assembly Routines to implement SGX Instructions
            1. EACCEPT
            2. EACCEPTCOPY
            3. EMODPE
        2. Low Level: APIs that are presented to the enclave to manage memory. These will include OCalls to untrusted host side to make calls to the kernel functions.
            1. Allocate Pages - similar to mmap(.., MAP_ANONYMOUS, ..) and/or VirtualAlloc().
                1. Include APIs to free pages.
                2. Potentially need options to eagerly/lazily EACCEPT and an EACCEPTCOPY option
            2. Change Page Protection/Type - similar to mprotect() and/or VirtualProtect()
    3. Configuration: How is an enclave configured for core design
        1. Enclave range - the virtual size of the enclave.
4. EDMM Enabled Features - enabled by the Low Level Functions which is part of core design. This should also include configuration and design of high level functions. These do not need to be completed at the same time but should be considered to make sure that the Core Design can support these (and more features that EDMM can enable):
    1. Reserving/Committing Memory outside of the heap (provided in Lower Level APIs) - similar to mmap()/munmap()
    2. Changing page protections (provided in Lower Level APIs) - similar to mprotect()
    3. Heap Expansion/Contraction - may need sbrk() implementation (instead of mmap)
    4. Stack Expansion - may be complicated since this must be done without a stack
    5. Thread Creation/Deletion : somewhat complex because this must create/initialize many pages
        1. Must add pages, change page type (for TCS), etc.
    6. Dynamic Code Loading